rust/types/src/spann_configuration.rs

```diff
@@ -0,0 +1,133 @@
+use crate::HnswSpace;
+use chroma_error::{ChromaError, ErrorCodes};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use utoipa::ToSchema;
+use validator::Validate;
+
+fn default_search_nprobe() -> u32 {
+    128
+}
+
+fn default_search_rng_factor() -> f32 {
+    1.0
+}
+
+fn default_search_rng_epsilon() -> f32 {
+    10.0
+}
+
+fn default_write_nprobe() -> u32 {
+    128
+}
+
+fn default_write_rng_factor() -> f32 {
+    1.0
+}
+
+fn default_write_rng_epsilon() -> f32 {
+    10.0
+}
+
+fn default_split_threshold() -> u32 {
+    100
+}
+
+fn default_num_samples_kmeans() -> usize {
+    1000
+}
+
+fn default_initial_lambda() -> f32 {
+    100.0
+}
+
+fn default_reassign_nbr_count() -> u32 {
+    8
+}
+
+fn default_merge_threshold() -> u32 {
+    50
+}
+
+fn default_num_centers_to_merge_to() -> u32 {
+    8
+}
+
+fn default_construction_ef_spann() -> usize {
+    200
+}
+
+fn default_search_ef_spann() -> usize {
+    200
+}
+
+fn default_m_spann() -> usize {
+    16
+}
+
+#[derive(Debug, Error)]
+pub enum DistributedSpannParametersFromSegmentError {
+    #[error("Invalid metadata: {0}")]
+    InvalidMetadata(#[from] serde_json::Error),
+    #[error("Invalid parameters: {0}")]
+    InvalidParameters(#[from] validator::ValidationErrors),
+}
+
+impl ChromaError for DistributedSpannParametersFromSegmentError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            DistributedSpannParametersFromSegmentError::InvalidMetadata(_) => {
+                ErrorCodes::InvalidArgument
+            }
+            DistributedSpannParametersFromSegmentError::InvalidParameters(_) => {
+                ErrorCodes::InvalidArgument
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Validate, PartialEq, ToSchema)]
+pub struct SpannConfiguration {
+    #[serde(default = "default_search_nprobe")]
+    #[validate(range(min = 8))]
+    pub search_nprobe: u32,
```

Comment:
cc @sanketkedia this will break any of your collections in staging configured with SPANN parameters